File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Java/17-分账/AddReceiver.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/partner/4014985777

import com.google.gson.annotations.SerializedName;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;

/**
 * 添加分账接收方（服务商模式）
 */
public class AddReceiver {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "POST";
  private static String PATH = "/v3/profitsharing/receivers/add";

  public static void main(String[] args) {
    AddReceiver client = new AddReceiver(
      "YOUR_MCHID",
      "YOUR_CERT_SERIAL_NO",
      "/path/to/apiclient_key.pem",
      "YOUR_PUB_KEY_ID",
      "/path/to/wxp_pub.pem"
    );

    AddReceiverRequest request = new AddReceiverRequest();
    request.subMchid = "1900000109";
    request.appid = "wxYOUR_APPID";
    request.subAppid = "wx8888888888888889";
    request.type = ReceiverType.MERCHANT_ID;
    request.account = "86693852";
    request.name = client.encrypt("name");
    request.relationType = ReceiverRelationType.STORE;
    request.customRelation = "代理商";
    try {
      AddReceiverResponse response = client.run(request);
      System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
      e.printStackTrace();
    }
  }

  public AddReceiverResponse run(AddReceiverRequest request) {
    String uri = PATH;
    String reqBody = WXPayUtility.toJson(request);

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo, privateKey, METHOD, uri, reqBody));
    reqBuilder.addHeader("Content-Type", "application/json");
    RequestBody requestBody = RequestBody.create(MediaType.parse("application/json; charset=utf-8"), reqBody);
    reqBuilder.method(METHOD, requestBody);
    Request httpRequest = reqBuilder.build();

    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey,
            httpResponse.headers(), respBody);
        return WXPayUtility.fromJson(respBody, AddReceiverResponse.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public AddReceiver(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public String encrypt(String plainText) {
    return WXPayUtility.encrypt(this.wechatPayPublicKey, plainText);
  }

  public static class AddReceiverRequest {
    @SerializedName("sub_mchid")
    public String subMchid;

    @SerializedName("appid")
    public String appid;

    @SerializedName("sub_appid")
    public String subAppid;

    @SerializedName("type")
    public ReceiverType type;

    @SerializedName("account")
    public String account;

    @SerializedName("name")
    public String name;

    @SerializedName("relation_type")
    public ReceiverRelationType relationType;

    @SerializedName("custom_relation")
    public String customRelation;
  }

  public static class AddReceiverResponse {
    @SerializedName("type")
    public ReceiverType type;

    @SerializedName("account")
    public String account;

    @SerializedName("name")
    public String name;

    @SerializedName("relation_type")
    public ReceiverRelationType relationType;

    @SerializedName("custom_relation")
    public String customRelation;
  }

  public enum ReceiverType {
    @SerializedName("MERCHANT_ID")
    MERCHANT_ID,
    @SerializedName("PERSONAL_OPENID")
    PERSONAL_OPENID,
    @SerializedName("PERSONAL_SUB_OPENID")
    PERSONAL_SUB_OPENID
  }

  public enum ReceiverRelationType {
    @SerializedName("STORE")
    STORE,
    @SerializedName("STAFF")
    STAFF,
    @SerializedName("STORE_OWNER")
    STORE_OWNER,
    @SerializedName("PARTNER")
    PARTNER,
    @SerializedName("HEADQUARTER")
    HEADQUARTER,
    @SerializedName("BRAND")
    BRAND,
    @SerializedName("DISTRIBUTOR")
    DISTRIBUTOR,
    @SerializedName("USER")
    USER,
    @SerializedName("SUPPLIER")
    SUPPLIER,
    @SerializedName("CUSTOM")
    CUSTOM,
    @SerializedName("SERVICE_PROVIDER")
    SERVICE_PROVIDER
  }

}

File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Go/17-分账/AddReceiver.go
package main

import (
	"bytes"
	"demo/wxpay_utility" // 引用微信支付工具库，参考 https://pay.weixin.qq.com/doc/v3/partner/4015119446
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
)

func main() {
	config, err := wxpay_utility.CreateMchConfig(
		"YOUR_MCHID",
		"YOUR_CERT_SERIAL_NO",
		"/path/to/apiclient_key.pem",
		"YOUR_PUB_KEY_ID",
		"/path/to/wxp_pub.pem",
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &AddReceiverRequest{
		SubMchid:       wxpay_utility.String("1900000109"),
		Appid:          wxpay_utility.String("wxYOUR_APPID"),
		SubAppid:       wxpay_utility.String("wx8888888888888889"),
		Type:           RECEIVERTYPE_MERCHANT_ID.Ptr(),
		Account:        wxpay_utility.String("86693852"),
		Name:           wxpay_utility.String("hu89ohu89ohu89o"), /*请传入wxpay_utility.EncryptOAEPWithPublicKey 加密结果*/
		RelationType:   RECEIVERRELATIONTYPE_STORE.Ptr(),
		CustomRelation: wxpay_utility.String("代理商"),
	}

	response, err := AddReceiver(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		return
	}

	fmt.Printf("请求成功: %+v\n", response)
}

func AddReceiver(config *wxpay_utility.MchConfig, request *AddReceiverRequest) (response *AddReceiverResponse, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "POST"
		path   = "/v3/profitsharing/receivers/add"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	reqBody, err := json.Marshal(request)
	if err != nil {
		return nil, err
	}
	httpRequest, err := http.NewRequest(method, reqUrl.String(), bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	httpRequest.Header.Set("Content-Type", "application/json")
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), reqBody)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		err = wxpay_utility.ValidateResponse(config.WechatPayPublicKeyId(), config.WechatPayPublicKey(), &httpResponse.Header, respBody)
		if err != nil {
			return nil, err
		}
		response := &AddReceiverResponse{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}
		return response, nil
	} else {
		return nil, wxpay_utility.NewApiException(httpResponse.StatusCode, httpResponse.Header, respBody)
	}
}

type AddReceiverRequest struct {
	SubMchid       *string               `json:"sub_mchid,omitempty"`
	Appid          *string               `json:"appid,omitempty"`
	SubAppid       *string               `json:"sub_appid,omitempty"`
	Type           *ReceiverType         `json:"type,omitempty"`
	Account        *string               `json:"account,omitempty"`
	Name           *string               `json:"name,omitempty"`
	RelationType   *ReceiverRelationType `json:"relation_type,omitempty"`
	CustomRelation *string               `json:"custom_relation,omitempty"`
}

type AddReceiverResponse struct {
	Type           *ReceiverType         `json:"type,omitempty"`
	Account        *string               `json:"account,omitempty"`
	Name           *string               `json:"name,omitempty"`
	RelationType   *ReceiverRelationType `json:"relation_type,omitempty"`
	CustomRelation *string               `json:"custom_relation,omitempty"`
}

type ReceiverType string

func (e ReceiverType) Ptr() *ReceiverType { return &e }

const (
	RECEIVERTYPE_MERCHANT_ID         ReceiverType = "MERCHANT_ID"
	RECEIVERTYPE_PERSONAL_OPENID     ReceiverType = "PERSONAL_OPENID"
	RECEIVERTYPE_PERSONAL_SUB_OPENID ReceiverType = "PERSONAL_SUB_OPENID"
)

type ReceiverRelationType string

func (e ReceiverRelationType) Ptr() *ReceiverRelationType { return &e }

const (
	RECEIVERRELATIONTYPE_STORE            ReceiverRelationType = "STORE"
	RECEIVERRELATIONTYPE_STAFF            ReceiverRelationType = "STAFF"
	RECEIVERRELATIONTYPE_STORE_OWNER      ReceiverRelationType = "STORE_OWNER"
	RECEIVERRELATIONTYPE_PARTNER          ReceiverRelationType = "PARTNER"
	RECEIVERRELATIONTYPE_HEADQUARTER      ReceiverRelationType = "HEADQUARTER"
	RECEIVERRELATIONTYPE_BRAND            ReceiverRelationType = "BRAND"
	RECEIVERRELATIONTYPE_DISTRIBUTOR      ReceiverRelationType = "DISTRIBUTOR"
	RECEIVERRELATIONTYPE_USER             ReceiverRelationType = "USER"
	RECEIVERRELATIONTYPE_SUPPLIER         ReceiverRelationType = "SUPPLIER"
	RECEIVERRELATIONTYPE_CUSTOM           ReceiverRelationType = "CUSTOM"
	RECEIVERRELATIONTYPE_SERVICE_PROVIDER ReceiverRelationType = "SERVICE_PROVIDER"
)
